File: torch_conversions/torch_conversions_py_cuda_plugin/test/test_torch_conversions_cuda.py
# Copyright 2026 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from importlib.metadata import entry_points
import subprocess
import sys

import pytest

import torch

import torch_conversions
from torch_conversions import allocate_tensor_msg
from torch_conversions import from_input_tensor_msg
from torch_conversions import from_output_tensor_msg
from torch_conversions import set_stream
from torch_conversions import to_tensor_msg


CUDA_AVAILABLE = torch_conversions._adapter_available('cuda')


def test_cuda_adapter_entry_point_is_discoverable():
    adapters = {
        entry_point.name: entry_point.value
        for entry_point in entry_points(group='torch_conversions.adapters')
    }

    assert adapters['cuda'] == 'torch_conversions_cuda_plugin:register'


def test_plugin_can_be_imported_before_core():
    subprocess.run(
        [
            sys.executable,
            '-c',
            'import torch_conversions_cuda_plugin; import torch_conversions',
        ],
        check=True,
    )


@pytest.mark.skipif(not CUDA_AVAILABLE, reason='CUDA support is unavailable')
def test_cuda_write_read_round_trip():
    msg = allocate_tensor_msg((6,), torch.float32, 'cuda')
    output = from_output_tensor_msg(msg)
    output.copy_(torch.arange(6, dtype=torch.float32, device='cuda'))

    result = from_input_tensor_msg(msg)

    assert result.is_cuda
    assert torch.equal(result.cpu(), torch.arange(6, dtype=torch.float32))


@pytest.mark.skipif(not CUDA_AVAILABLE, reason='CUDA support is unavailable')
def test_cuda_tensor_to_message_round_trip():
    source = torch.arange(12, dtype=torch.int32, device='cuda').reshape(3, 4)

    msg = to_tensor_msg(source)
    result = from_input_tensor_msg(msg, clone=False)

    assert msg.data.backend_type == 'cuda'
    assert result.is_cuda
    assert torch.equal(result.cpu(), source.cpu())


@pytest.mark.skipif(not CUDA_AVAILABLE, reason='CUDA support is unavailable')
def test_cuda_zero_copy_view_aliases_message_storage():
    msg = allocate_tensor_msg((8,), torch.float32, 'cuda')
    output = from_output_tensor_msg(msg)
    output.copy_(torch.arange(8, dtype=torch.float32, device='cuda'))
    output_pointer = output.data_ptr()
    del output

    view = from_input_tensor_msg(msg, clone=False)

    assert view.data_ptr() == output_pointer
    assert torch.equal(view.cpu(), torch.arange(8, dtype=torch.float32))


@pytest.mark.skipif(not CUDA_AVAILABLE, reason='CUDA support is unavailable')
def test_set_stream_selects_non_default_cuda_stream():
    default_stream = torch.cuda.current_stream().cuda_stream

    with set_stream('cuda'):
        selected_stream = torch.cuda.current_stream().cuda_stream
        msg = allocate_tensor_msg((1,), torch.float32, 'cuda')
        output = from_output_tensor_msg(msg)
        output.fill_(1)

    assert selected_stream != default_stream

File: torch_conversions/torch_conversions/test/test_torch_conversions_cuda.cpp
// Copyright 2026 Open Source Robotics Foundation, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <gtest/gtest.h>
#include <torch/torch.h>

#include <c10/cuda/CUDAGuard.h>
#include <c10/cuda/CUDAStream.h>

#include <chrono>
#include <stdexcept>
#include <thread>

#include "torch_conversions/torch_conversions.hpp"

namespace
{

void * current_stream()
{
  return c10::cuda::getCurrentCUDAStream().stream();
}

}  // namespace

TEST(TorchConversionsCuda, CloneWaitsOnTheSuppliedConsumerStream)
{
  if (!torch::cuda::is_available()) {
    GTEST_SKIP() << "CUDA driver is unavailable";
  }
  const auto producer = c10::cuda::getStreamFromPool();
  const auto consumer = c10::cuda::getStreamFromPool();
  auto msg = torch_conversions::allocate_tensor_msg({4}, at::kFloat, c10::kCUDA);
  {
    c10::cuda::CUDAStreamGuard guard(producer);
    auto output = torch_conversions::from_output_tensor_msg(*msg, producer.stream());
    output.zero_();
    producer.synchronize();
    ASSERT_EQ(cudaLaunchHostFunc(producer.stream(), [](void *) {
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
      }, nullptr), cudaSuccess);
    output.fill_(7);
  }
  auto clone = torch_conversions::from_input_tensor_msg(*msg, true, consumer.stream());
  consumer.synchronize();
  EXPECT_TRUE(torch::equal(clone.cpu(), torch::full({4}, 7.0)));
}

TEST(TorchConversionsCuda, NonContiguousCopyUsesTheSuppliedStream)
{
  if (!torch::cuda::is_available()) {
    GTEST_SKIP() << "CUDA driver is unavailable";
  }
  const auto stream = c10::cuda::getStreamFromPool();
  at::Tensor source;
  {
    c10::cuda::CUDAStreamGuard guard(stream);
    source = torch::zeros({2, 3}, at::device(at::kCUDA).dtype(at::kFloat));
    stream.synchronize();
    ASSERT_EQ(cudaLaunchHostFunc(stream.stream(), [](void *) {
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
      }, nullptr), cudaSuccess);
    source.fill_(5);
  }
  auto msg = torch_conversions::to_tensor_msg(source.transpose(0, 1), stream.stream());
  source = at::Tensor{};
  auto value = torch_conversions::from_input_tensor_msg(*msg, true, stream.stream());
  stream.synchronize();
  EXPECT_TRUE(torch::equal(value.cpu(), torch::full({3, 2}, 5.0)));
}

TEST(TorchConversionsCuda, InputCloneSurvivesMessageDestruction)
{
  if (!torch::cuda::is_available()) {
    GTEST_SKIP() << "CUDA driver is unavailable";
  }
  const auto stream = c10::cuda::getStreamFromPool();
  c10::cuda::CUDAStreamGuard guard(stream);
  auto msg = torch_conversions::allocate_tensor_msg({4}, at::kFloat, c10::kCUDA);
  auto output = torch_conversions::from_output_tensor_msg(*msg, stream.stream());
  auto * pointer = output.data_ptr();
  output.fill_(9);
  output = at::Tensor{};
  auto input = torch_conversions::from_input_tensor_msg(*msg, true, stream.stream());
  msg.reset();
  EXPECT_NE(input.data_ptr(), pointer);
  EXPECT_TRUE(torch::equal(input.cpu(), torch::full({4}, 9.0)));
}

TEST(TorchConversionsCuda, PreservesAndValidatesDeviceIndices)
{
  if (!torch::cuda::is_available()) {
    GTEST_SKIP() << "CUDA driver is unavailable";
  }
  int count = 0;
  ASSERT_EQ(cudaGetDeviceCount(&count), cudaSuccess);
  EXPECT_THROW(torch_conversions::set_stream(c10::Device(c10::kCUDA, count)), c10::Error);
  EXPECT_THROW(torch_conversions::allocate_tensor_msg(
      {1}, at::kFloat, c10::Device(c10::kCUDA, count)), c10::Error);
  const int current = c10::cuda::current_device();
  for (int device = 0; device < count; ++device) {
    if (device != current) {
      EXPECT_THROW(torch_conversions::allocate_tensor_msg(
          {1}, at::kFloat, c10::Device(c10::kCUDA, device)), std::runtime_error);
      continue;
    }
    auto msg = torch_conversions::allocate_tensor_msg(
      {1}, at::kFloat, c10::Device(c10::kCUDA, device));
    auto value = torch_conversions::from_output_tensor_msg(*msg);
    EXPECT_EQ(value.device().index(), device);
  }
}

TEST(TorchConversionsCuda, AllocatesCudaStorage)
{
  if (!torch::cuda::is_available()) {
    GTEST_SKIP() << "CUDA driver is unavailable";
  }

  auto msg = torch_conversions::allocate_tensor_msg(
    {4}, at::kFloat, c10::kCUDA);
  EXPECT_EQ(msg->data.get_backend_type(), "cuda");

  auto tensor = torch_conversions::from_output_tensor_msg(*msg);
  EXPECT_TRUE(tensor.is_cuda());
  tensor.fill_(1.0);
}

TEST(TorchConversionsCuda, RoundTripsOnACallerSuppliedStream)
{
  if (!torch::cuda::is_available()) {
    GTEST_SKIP() << "CUDA driver is unavailable";
  }

  auto msg = torch_conversions::allocate_tensor_msg(
    {64}, at::kFloat, c10::kCUDA);

  {
    auto output = torch_conversions::from_output_tensor_msg(
      *msg, current_stream());
    ASSERT_TRUE(output.is_cuda());
    output.fill_(3.0);
  }

  auto input = torch_conversions::from_input_tensor_msg(
    *msg, /*clone=*/true, current_stream());
  ASSERT_TRUE(input.is_cuda());
  EXPECT_FLOAT_EQ(input.sum().item<float>(), 192.0f);
}

TEST(TorchConversionsCuda, CopiesIntoStorageOnACallerSuppliedStream)
{
  if (!torch::cuda::is_available()) {
    GTEST_SKIP() << "CUDA driver is unavailable";
  }

  const auto source = at::full({32}, 5.0, at::device(at::kCUDA).dtype(at::kFloat));

  auto msg = torch_conversions::to_tensor_msg(source, current_stream());
  ASSERT_EQ(msg->data.get_backend_type(), "cuda");

  auto readback = torch_conversions::from_input_tensor_msg(
    *msg, /*clone=*/true, current_stream());
  EXPECT_FLOAT_EQ(readback.sum().item<float>(), 160.0f);
}

TEST(TorchConversionsCuda, ScopedStreamRestoresState)
{
  if (!torch::cuda::is_available()) {
    GTEST_SKIP() << "CUDA driver is unavailable";
  }

  const auto original = c10::cuda::getCurrentCUDAStream();
  {
    auto guard = torch_conversions::set_stream();
    EXPECT_EQ(
      c10::cuda::getCurrentCUDAStream() != original,
      torch_conversions::default_backend() == "cuda");
  }
  EXPECT_EQ(c10::cuda::getCurrentCUDAStream(), original);
  {
    auto guard = torch_conversions::set_stream(original.device());
    const auto stream = c10::cuda::getCurrentCUDAStream();
    EXPECT_NE(stream, original);
    EXPECT_NE(stream, c10::cuda::getDefaultCUDAStream());
    {
      auto cpu_guard = torch_conversions::set_stream(c10::kCPU);
      EXPECT_EQ(c10::cuda::getCurrentCUDAStream(), stream);
    }
    try {
      torch_conversions::StreamGuard nested(original.device());
      EXPECT_NE(c10::cuda::getCurrentCUDAStream(), stream);
      throw std::runtime_error("unwind stream guard");
    } catch (const std::runtime_error &) {
      EXPECT_EQ(c10::cuda::getCurrentCUDAStream(), stream);
    }

    auto msg = torch_conversions::allocate_tensor_msg({32}, at::kFloat, original.device());
    {
      auto output = torch_conversions::from_output_tensor_msg(*msg);
      output.fill_(7.0);
    }
    auto input = torch_conversions::from_input_tensor_msg(*msg);
    EXPECT_FLOAT_EQ(input.sum().item<float>(), 224.0f);
  }
  EXPECT_EQ(c10::cuda::getCurrentCUDAStream(), original);
  EXPECT_EQ(c10::cuda::current_device(), original.device_index());
}
